dataRetriever/dataPool/proofsCache/proofsCache.go

```diff
@@ -43,39 +46,76 @@ func (pc *proofsCache) addProof(proof data.HeaderProofHandler) {
 		return
 	}
 
-	pc.mutProofsCache.Lock()
-	defer pc.mutProofsCache.Unlock()
+	pc.insertProofByNonce(proof)
 
-	pc.proofsByNonce = append(pc.proofsByNonce, &proofNonceMapping{
-		headerHash: string(proof.GetHeaderHash()),
-		nonce:      proof.GetHeaderNonce(),
-	})
+	pc.mutProofsByHash.Lock()
+	pc.proofsByHash[string(proof.GetHeaderHash())] = proof
+	pc.mutProofsByHash.Unlock()
+}
 
-	sort.Slice(pc.proofsByNonce, func(i, j int) bool {
-		return pc.proofsByNonce[i].nonce < pc.proofsByNonce[j].nonce
-	})
+func (pc *proofsCache) insertProofByNonce(proof data.HeaderProofHandler) {
+	pc.mutProofsByNonce.Lock()
+	defer pc.mutProofsByNonce.Unlock()
 
-	pc.proofsByHash[string(proof.GetHeaderHash())] = proof
+	if len(pc.proofsByNonceBuckets) == 0 {
+		pc.insertInNewBucket(proof)
+		return
+	}
+
+	headBucket := pc.proofsByNonceBuckets[0]
+
+	if headBucket.isFull() {
+		pc.insertInNewBucket(proof)
+		return
+	}
+
+	headBucket.insertInExisting(proof)
+}
+
+func (pc *proofsCache) insertInNewBucket(proof data.HeaderProofHandler) {
+	bucket := newProofBucket(pc.bucketSize)
+	bucket.insertInNew(proof)
+
+	pc.proofsByNonceBuckets = append([]*proofNonceBucket{bucket}, pc.proofsByNonceBuckets...)
 }
 
 func (pc *proofsCache) cleanupProofsBehindNonce(nonce uint64) {
 	if nonce == 0 {
 		return
 	}
 
-	pc.mutProofsCache.Lock()
-	defer pc.mutProofsCache.Unlock()
+	pc.mutProofsByNonce.Lock()
+	defer pc.mutProofsByNonce.Unlock()
+
+	buckets := make([]*proofNonceBucket, 0)
+
+	wg := &sync.WaitGroup{}
+
+	for _, bucket := range pc.proofsByNonceBuckets {
+		if nonce > bucket.maxNonce {
+			wg.Add(1)
 
-	proofsByNonce := make([]*proofNonceMapping, 0)
+			go func(bucket *proofNonceBucket) {
+				pc.cleanupProofsInBucket(bucket)
+				wg.Done()
+			}(bucket)
 
-	for _, proofInfo := range pc.proofsByNonce {
-		if proofInfo.nonce < nonce {
-			delete(pc.proofsByHash, proofInfo.headerHash)
 			continue
 		}
 
-		proofsByNonce = append(proofsByNonce, proofInfo)
+		buckets = append(buckets, bucket)
 	}
 
-	pc.proofsByNonce = proofsByNonce
+	wg.Wait()
+
+	pc.proofsByNonceBuckets = buckets
+}
+
+func (pc *proofsCache) cleanupProofsInBucket(bucket *proofNonceBucket) {
+	pc.mutProofsByHash.Lock()
+	defer pc.mutProofsByHash.Unlock()
```

Comment:
The caller function also locks (sorry if I'm mistaken).

Reply:
there 2 separate locks

Reply:
Indeed, my bad / overlooked!